File: output/cpp/tests/13_system_return.cpp
#include <string>
#include <unordered_map>
#include <vector>
#include <any>
#include <memory>


#include <iostream>
#include <string>
#include <cassert>

class SystemReturnTestFrameEvent {
public:
    std::string _message;
    std::unordered_map<std::string, std::any> _parameters;

    SystemReturnTestFrameEvent(const std::string& message, std::unordered_map<std::string, std::any> params = {})
        : _message(message), _parameters(std::move(params)) {}
};

class SystemReturnTestFrameContext {
public:
    SystemReturnTestFrameEvent _event;
    std::any _return;
    std::unordered_map<std::string, std::any> _data;

    SystemReturnTestFrameContext(SystemReturnTestFrameEvent event, std::any default_return = {})
        : _event(std::move(event)), _return(std::move(default_return)) {}
};

class SystemReturnTestCompartment {
public:
    std::string state;
    std::unordered_map<std::string, std::any> state_args;
    std::unordered_map<std::string, std::any> state_vars;
    std::unordered_map<std::string, std::any> enter_args;
    std::unordered_map<std::string, std::any> exit_args;
    std::unique_ptr<SystemReturnTestFrameEvent> forward_event;
    std::unique_ptr<SystemReturnTestCompartment> parent_compartment;

    explicit SystemReturnTestCompartment(const std::string& state) : state(state) {}

    std::unique_ptr<SystemReturnTestCompartment> clone() const {
        auto c = std::make_unique<SystemReturnTestCompartment>(state);
        c->state_args = state_args;
        c->state_vars = state_vars;
        c->enter_args = enter_args;
        c->exit_args = exit_args;
        return c;
    }
};

class SystemReturnTest {
private:
    std::vector<std::unique_ptr<SystemReturnTestCompartment>> _state_stack;
    std::unique_ptr<SystemReturnTestCompartment> __compartment;
    std::unique_ptr<SystemReturnTestCompartment> __next_compartment;
    std::vector<SystemReturnTestFrameContext> _context_stack;

    void __kernel(SystemReturnTestFrameEvent& __e) {
        __router(__e);
        while (__next_compartment) {
            auto next_compartment = std::move(__next_compartment);
            SystemReturnTestFrameEvent exit_event("<$");
            __router(exit_event);
            __compartment = std::move(next_compartment);
            if (!__compartment->forward_event) {
                SystemReturnTestFrameEvent enter_event("$>");
                __router(enter_event);
            } else {
                auto forward_event = std::move(__compartment->forward_event);
                if (forward_event->_message == "$>") {
                    __router(*forward_event);
                } else {
                    SystemReturnTestFrameEvent enter_event("$>");
                    __router(enter_event);
                    __router(*forward_event);
                }
            }
        }
    }

    void __router(SystemReturnTestFrameEvent& __e) {
        const auto& state_name = __compartment->state;
        if (state_name == "Calculator") {
            _state_Calculator(__e);
        }
    }

    void __transition(std::unique_ptr<SystemReturnTestCompartment> next) {
        __next_compartment = std::move(next);
    }

    void _state_Calculator(SystemReturnTestFrameEvent& __e) {
        auto* __sv_comp = __compartment.get();
        while (__sv_comp && __sv_comp->state != "Calculator") { __sv_comp = __sv_comp->parent_compartment.get(); }
        if (__e._message == "$>") {
            if (__compartment->state_vars.count("value") == 0) { __compartment->state_vars["value"] = std::any(0); }
        } else if (__e._message == "add") {
            auto a = std::any_cast<int>(__e._parameters.at("a"));
            auto b = std::any_cast<int>(__e._parameters.at("b"));
            _context_stack.back()._return = std::any(a + b);
            return;;
        } else if (__e._message == "get_value") {
            __sv_comp->state_vars["value"] = std::any(42);
            _context_stack.back()._return = std::any(std::any_cast<int>(__sv_comp->state_vars["value"]));
            return;;
        } else if (__e._message == "greet") {
            auto name = std::any_cast<std::string>(__e._parameters.at("name"));
            _context_stack.back()._return = std::any(std::string("Hello, ") + name + "!");
            return;;
        } else if (__e._message == "multiply") {
            auto a = std::any_cast<int>(__e._parameters.at("a"));
            auto b = std::any_cast<int>(__e._parameters.at("b"));
            _context_stack.back()._return = std::any(a * b);
        }
    }

public:
    SystemReturnTest() {
        __compartment = std::make_unique<SystemReturnTestCompartment>("Calculator");
        SystemReturnTestFrameEvent __frame_event("$>");
        SystemReturnTestFrameContext __ctx(std::move(__frame_event));
        _context_stack.push_back(std::move(__ctx));
        __kernel(_context_stack.back()._event);
        _context_stack.pop_back();
    }

    int add(int a, int b) {
        std::unordered_map<std::string, std::any> __params;
        __params["a"] = a;
        __params["b"] = b;
        SystemReturnTestFrameEvent __e("add", std::move(__params));
        SystemReturnTestFrameContext __ctx(std::move(__e), std::any(int()));
        _context_stack.push_back(std::move(__ctx));
        __kernel(_context_stack.back()._event);
        auto __result = std::any_cast<int>(std::move(_context_stack.back()._return));
        _context_stack.pop_back();
        return __result;
    }

    int multiply(int a, int b) {
        std::unordered_map<std::string, std::any> __params;
        __params["a"] = a;
        __params["b"] = b;
        SystemReturnTestFrameEvent __e("multiply", std::move(__params));
        SystemReturnTestFrameContext __ctx(std::move(__e), std::any(int()));
        _context_stack.push_back(std::move(__ctx));
        __kernel(_context_stack.back()._event);
        auto __result = std::any_cast<int>(std::move(_context_stack.back()._return));
        _context_stack.pop_back();
        return __result;
    }

    std::string greet(std::string name) {
        std::unordered_map<std::string, std::any> __params;
        __params["name"] = name;
        SystemReturnTestFrameEvent __e("greet", std::move(__params));
        SystemReturnTestFrameContext __ctx(std::move(__e), std::any(std::string()));
        _context_stack.push_back(std::move(__ctx));
        __kernel(_context_stack.back()._event);
        auto __result = std::any_cast<std::string>(std::move(_context_stack.back()._return));
        _context_stack.pop_back();
        return __result;
    }

    int get_value() {
        SystemReturnTestFrameEvent __e("get_value");
        SystemReturnTestFrameContext __ctx(std::move(__e), std::any(int()));
        _context_stack.push_back(std::move(__ctx));
        __kernel(_context_stack.back()._event);
        auto __result = std::any_cast<int>(std::move(_context_stack.back()._return));
        _context_stack.pop_back();
        return __result;
    }
};

int main() {
    printf("=== Test 13: System Return ===\n");
    SystemReturnTest calc;

    int result = calc.add(3, 5);
    if (result != 8) {
        printf("FAIL: Expected 8, got %d\n", result);
        assert(false);
    }
    printf("add(3, 5) = %d\n", result);

    result = calc.multiply(4, 7);
    if (result != 28) {
        printf("FAIL: Expected 28, got %d\n", result);
        assert(false);
    }
    printf("multiply(4, 7) = %d\n", result);

    std::string greeting = calc.greet("World");
    if (greeting != "Hello, World!") {
        printf("FAIL: Expected 'Hello, World!', got '%s'\n", greeting.c_str());
        assert(false);
    }
    printf("greet('World') = %s\n", greeting.c_str());

    int value = calc.get_value();
    if (value != 42) {
        printf("FAIL: Expected 42, got %d\n", value);
        assert(false);
    }
    printf("get_value() = %d\n", value);

    printf("PASS: System return works correctly\n");
    return 0;
}

File: output/go/tests/13_system_return.go

package main

import "fmt"
import "os"

type SystemReturnTestFrameEvent struct {
    _message    string
    _parameters map[string]any
}

type SystemReturnTestFrameContext struct {
    _event  SystemReturnTestFrameEvent
    _return any
    _data   map[string]any
}

type SystemReturnTestCompartment struct {
    state            string
    stateArgs        map[string]any
    stateVars        map[string]any
    enterArgs        map[string]any
    exitArgs         map[string]any
    forwardEvent     *SystemReturnTestFrameEvent
    parentCompartment *SystemReturnTestCompartment
}

func newSystemReturnTestCompartment(state string) *SystemReturnTestCompartment {
    return &SystemReturnTestCompartment{
        state:    state,
        stateArgs: make(map[string]any),
        stateVars: make(map[string]any),
        enterArgs: make(map[string]any),
        exitArgs:  make(map[string]any),
    }
}

func (c *SystemReturnTestCompartment) copy() *SystemReturnTestCompartment {
    nc := &SystemReturnTestCompartment{
        state: c.state,
        stateArgs: make(map[string]any),
        stateVars: make(map[string]any),
        enterArgs: make(map[string]any),
        exitArgs:  make(map[string]any),
        forwardEvent:     c.forwardEvent,
        parentCompartment: c.parentCompartment,
    }
    for k, v := range c.stateArgs { nc.stateArgs[k] = v }
    for k, v := range c.stateVars { nc.stateVars[k] = v }
    for k, v := range c.enterArgs { nc.enterArgs[k] = v }
    for k, v := range c.exitArgs { nc.exitArgs[k] = v }
    return nc
}

type SystemReturnTest struct {
    _state_stack []*SystemReturnTestCompartment
    __compartment *SystemReturnTestCompartment
    __next_compartment *SystemReturnTestCompartment
    _context_stack []SystemReturnTestFrameContext
}

func NewSystemReturnTest() *SystemReturnTest {
    s := &SystemReturnTest{}
    s._state_stack = make([]*SystemReturnTestCompartment, 0)
    s._context_stack = make([]SystemReturnTestFrameContext, 0)
    s.__compartment = newSystemReturnTestCompartment("Calculator")
    s.__next_compartment = nil
    __frame_event := SystemReturnTestFrameEvent{_message: "$>", _parameters: nil}
    __ctx := SystemReturnTestFrameContext{_event: __frame_event, _data: make(map[string]any)}
    s._context_stack = append(s._context_stack, __ctx)
    s.__kernel(&s._context_stack[len(s._context_stack)-1]._event)
    s._context_stack = s._context_stack[:len(s._context_stack)-1]
    return s
}

func (s *SystemReturnTest) __kernel(__e *SystemReturnTestFrameEvent) {
    s.__router(__e)
    for s.__next_compartment != nil {
        next_compartment := s.__next_compartment
        s.__next_compartment = nil
        exit_event := &SystemReturnTestFrameEvent{_message: "<$", _parameters: s.__compartment.exitArgs}
        s.__router(exit_event)
        s.__compartment = next_compartment
        if s.__compartment.forwardEvent == nil {
            enter_event := &SystemReturnTestFrameEvent{_message: "$>", _parameters: s.__compartment.enterArgs}
            s.__router(enter_event)
        } else {
            forward_event := s.__compartment.forwardEvent
            s.__compartment.forwardEvent = nil
            if forward_event._message == "$>" {
                s.__router(forward_event)
            } else {
                enter_event := &SystemReturnTestFrameEvent{_message: "$>", _parameters: s.__compartment.enterArgs}
                s.__router(enter_event)
                s.__router(forward_event)
            }
        }
    }
}

func (s *SystemReturnTest) __router(__e *SystemReturnTestFrameEvent) {
    switch s.__compartment.state {
    case "Calculator":
        s._state_Calculator(__e)
    }
}

func (s *SystemReturnTest) __transition(next *SystemReturnTestCompartment) {
    s.__next_compartment = next
}

func (s *SystemReturnTest) Add(a int, b int) int {
    __params := map[string]any{
        "a": a,
        "b": b,
    }
    __e := SystemReturnTestFrameEvent{_message: "Add", _parameters: __params}
    __ctx := SystemReturnTestFrameContext{_event: __e, _data: make(map[string]any)}
    s._context_stack = append(s._context_stack, __ctx)
    s.__kernel(&s._context_stack[len(s._context_stack)-1]._event)
    var __result int
    if __rv := s._context_stack[len(s._context_stack)-1]._return; __rv != nil { __result = __rv.(int) }
    s._context_stack = s._context_stack[:len(s._context_stack)-1]
    return __result
}

func (s *SystemReturnTest) Multiply(a int, b int) int {
    __params := map[string]any{
        "a": a,
        "b": b,
    }
    __e := SystemReturnTestFrameEvent{_message: "Multiply", _parameters: __params}
    __ctx := SystemReturnTestFrameContext{_event: __e, _data: make(map[string]any)}
    s._context_stack = append(s._context_stack, __ctx)
    s.__kernel(&s._context_stack[len(s._context_stack)-1]._event)
    var __result int
    if __rv := s._context_stack[len(s._context_stack)-1]._return; __rv != nil { __result = __rv.(int) }
    s._context_stack = s._context_stack[:len(s._context_stack)-1]
    return __result
}

func (s *SystemReturnTest) Greet(name string) string {
    __params := map[string]any{
        "name": name,
    }
    __e := SystemReturnTestFrameEvent{_message: "Greet", _parameters: __params}
    __ctx := SystemReturnTestFrameContext{_event: __e, _data: make(map[string]any)}
    s._context_stack = append(s._context_stack, __ctx)
    s.__kernel(&s._context_stack[len(s._context_stack)-1]._event)
    var __result string
    if __rv := s._context_stack[len(s._context_stack)-1]._return; __rv != nil { __result = __rv.(string) }
    s._context_stack = s._context_stack[:len(s._context_stack)-1]
    return __result
}

func (s *SystemReturnTest) GetValue() int {
    __e := SystemReturnTestFrameEvent{_message: "GetValue"}
    __ctx := SystemReturnTestFrameContext{_event: __e, _data: make(map[string]any)}
    s._context_stack = append(s._context_stack, __ctx)
    s.__kernel(&s._context_stack[len(s._context_stack)-1]._event)
    var __result int
    if __rv := s._context_stack[len(s._context_stack)-1]._return; __rv != nil { __result = __rv.(int) }
    s._context_stack = s._context_stack[:len(s._context_stack)-1]
    return __result
}

func (s *SystemReturnTest) _state_Calculator(__e *SystemReturnTestFrameEvent) {
    __sv_comp := s.__compartment
    for __sv_comp != nil && __sv_comp.state != "Calculator" { __sv_comp = __sv_comp.parentCompartment }
    if __e._message == "$>" {
        if _, ok := __sv_comp.stateVars["value"]; !ok {
            __sv_comp.stateVars["value"] = 0
        }
    } else if __e._message == "Add" {
        a := __e._parameters["a"].(int)
        _ = a
        b := __e._parameters["b"].(int)
        _ = b
        s._context_stack[len(s._context_stack)-1]._return = a + b
        return
    } else if __e._message == "GetValue" {
        __sv_comp.stateVars["value"] = 42
        s._context_stack[len(s._context_stack)-1]._return = __sv_comp.stateVars["value"].(int)
        return
    } else if __e._message == "Greet" {
        name := __e._parameters["name"].(string)
        _ = name
        s._context_stack[len(s._context_stack)-1]._return = "Hello, " + name + "!"
        return
    } else if __e._message == "Multiply" {
        a := __e._parameters["a"].(int)
        _ = a
        b := __e._parameters["b"].(int)
        _ = b
        s._context_stack[len(s._context_stack)-1]._return = a * b
    }
}

func main() {
	fmt.Println("=== Test 13: System Return ===")
	calc := NewSystemReturnTest()

	// Test return sugar
	result := calc.Add(3, 5)
	if result != 8 {
		fmt.Printf("FAIL: Expected 8, got %d\n", result)
		os.Exit(1)
	}
	fmt.Printf("Add(3, 5) = %d\n", result)

	// Test @@:return = expr
	result = calc.Multiply(4, 7)
	if result != 28 {
		fmt.Printf("FAIL: Expected 28, got %d\n", result)
		os.Exit(1)
	}
	fmt.Printf("Multiply(4, 7) = %d\n", result)

	// Test string return
	greeting := calc.Greet("World")
	if greeting != "Hello, World!" {
		fmt.Printf("FAIL: Expected 'Hello, World!', got '%s'\n", greeting)
		os.Exit(1)
	}
	fmt.Printf("Greet('World') = %s\n", greeting)

	// Test return with state variable
	value := calc.GetValue()
	if value != 42 {
		fmt.Printf("FAIL: Expected 42, got %d\n", value)
		os.Exit(1)
	}
	fmt.Printf("GetValue() = %d\n", value)

	fmt.Println("PASS: System return works correctly")
}
